### ISSUE TITLE
SIGHUP signal not raises exit immediately

### ISSUE THREAD

**Issue by FlyingWombat** [xonsh/xonsh#5381]
I'm not sure if this is an alacritty issue, or a xonsh issue, so I'm starting here.
When using alacritty in single-instance mode, xonsh processes remain if their parent window is closed by the window manager (e.g. Meta+Q for Sway). I've only been able to find this issue when using both alacritty in single-instance mode and xonsh. Alacritty single-instance with Zsh works as expected, alacritty individual instances works with xonsh, and xonsh with foot's single-instance mode works. The issue also does not occur if xonsh has a parent process other than alacritty (e.g. alacritty -> zsh -> xonsh)

## Expected Behavior

Xonsh processes should exit when parent window closes.

## Current Behavior

Xonsh process remains when parent window is closed.

## Steps to Reproduce

1. `export SHELL=$(which xonsh)` if xonsh isn't your default shell.
2. Start an alacritty instance with `alacritty --socket /tmp/alacritty.sock`
3. Create a second alacritty window with `alacritty msg --socket /tmp/alacritty.sock create-window`. This creates a new terminal attached to the same instance of alacritty. Notice now there are two xonsh processes attached to the same alacritty process.
4. (optional) run some command in the second terminal window to help identify it (e.g. man bash).
5. Close the second terminal window using the window manager instead of exiting xonsh directly. Notice there are still two xonsh processes attached to the alacritty process.

## xonfig

<details>

```
+------------------+----------------------------------+
| xonsh            | 0.16.0                           |
| Python           | 3.11.9                           |
| PLY              | 3.11                             |
| have readline    | True                             |
| prompt toolkit   | 3.0.40                           |
| shell type       | prompt_toolkit                   |
| history backend  | sqlite                           |
| pygments         | 2.17.2                           |
| on posix         | True                             |
| on linux         | True                             |
| distro           | unknown                          |
| on wsl           | False                            |
| on darwin        | False                            |
| on windows       | False                            |
| on cygwin        | False                            |
| on msys2         | False                            |
| is superuser     | False                            |
| default encoding | utf-8                            |
| xonsh encoding   | utf-8                            |
| encoding errors  | surrogateescape                  |
| xontrib 1        | vox                              |
| xontrib 2        | coreutils                        |
| xontrib 3        | voxapi                           |
| RC file 1        | /etc/xonsh/xonshrc               |
| RC file 2        | /home/wombat/.config/xonsh/rc.xsh |
+------------------+----------------------------------+
```

</details>

### System
Distro: Nixos
WM: Sway 1.9
Alacritty: 0.13.2

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by FlyingWombat** [comment]
> Please provide repeatable steps.

@anki-code You need to run `export SHELL=$(which xonsh)` first, before starting alacritty, in order for xonsh to be the default shell that alacritty spawns.

The steps to reproduce I described in OP consistently produce the issue on my system. Let me know if you still can't reproduce it on your system.

**Comment by FlyingWombat** [comment]
You skipped step 3 to create a second alacritty window. Then step 5 is to close that second window while leaving the first window open.

The idea here is to use a single alacritty instance to manage all terminal windows, instead of creating a new process for each window like normal. This is called "daemon mode" in some other terminal emulators. I have reproduced this also on Arch Linux.

Again, here are the steps to reproduce the issue:
```bash
# step 1, to make sure alacritty opens with xonsh as its direct child process
export SHELL=$(which xonsh)

# step 2, start alacritty
alacritty --socket /tmp/alacritty.sock
# alacritty terminal window opens with xonsh

# step 3, create second alacritty window attached to the same process; doesn't matter which terminal you run this from
alacritty msg --socket /tmp/alacritty.sock create-window
# a second alacritty window opens with xonsh

# step 4 (optional) just to help identify the unreachable process
man bash

# step 5 close the second alacritty window (the one with `man bash`) with WM action (red cross).
# Notice how there are still two xonsh processes; the one with `man bash` is still running.
```

-----------------

BTW, I tried using strace on the xonsh process, but it printed over 10000 lines of output and I have no idea what I'm looking for. Do you have any tips to help narrow it down?

**Comment by FlyingWombat** [comment]
>  Error: Custom { kind: ConnectionRefused, error: "invalid socket path \"/tmp/alacritty.sock\"" }

Strange. Maybe Mac has some weird rules about where sockets can be. The socket can be anywhere, I just used /tmp/ for an example.

Using the `signal-catch.py` example, with `export SHELL="$PWD/signal-catch.py"`, it appears that alacritty is sending a SIGHUP when the window is closed. But, that simple check doesn't show any different behavior between alacritty running normally (one process per window) and in single-instance (one process all windows); it just shows a single SIGHUP for ether one.

Turns out the strace logs are much shorter on Arch Linux (only 250 lines), so I'm attaching them here.

Regular alacritty (xonsh closes): 
[xonsh-strace-normal-closes.txt](https://github.com/xonsh/xonsh/files/15214280/xonsh-strace-normal-closes.txt)


single-instance (doesn't close):
[xonsh-strace-daemon-noclose.txt](https://github.com/xonsh/xonsh/files/15214281/xonsh-strace-daemon-noclose.txt)
